crates/ruff_python_parser/src/parser/expression.rs

```diff
@@ -1313,6 +1313,26 @@ impl<'src> Parser<'src> {
                 Expr::SetComp(self.parse_set_comprehension_expression(key_or_element.expr, start))
             }
             TokenKind::Colon => {
+                // Now, we know that it's either a dictionary expression or a dictionary comprehension.
+                // In either case, the key is limited to an `expression`.
+                if !key_or_element.is_parenthesized {
+                    match key_or_element.expr {
+                        Expr::Starred(_) => self.add_error(
+                            ParseErrorType::OtherError(
+                                "starred expression cannot be used here".to_string(),
+                            ),
+                            &key_or_element.expr,
+                        ),
```

Comment:
Even parenthesized starred expression isn't allowed but that check will be done by `parse_parenthesized_expression` and not here.